openrtb_ext/request.go

```diff
@@ -71,14 +71,186 @@ func (ert *ExtRequestTargeting) UnmarshalJSON(b []byte) error {
 }
 
 // PriceGranularity defines the allowed values for bidrequest.ext.prebid.targeting.pricegranularity
-type PriceGranularity string
-
-const (
-	PriceGranularityLow    PriceGranularity = "low"
-	PriceGranularityMedium PriceGranularity = "medium"
-	// Seems that PBS was written with medium = "med", so hacking that in
-	PriceGranularityMedPBS PriceGranularity = "med"
-	PriceGranularityHigh   PriceGranularity = "high"
-	PriceGranularityAuto   PriceGranularity = "auto"
-	PriceGranularityDense  PriceGranularity = "dense"
-)
+type PriceGranularity []GranularityRange
+
+type granularityRangeRaw struct {
+	Precision *int    `json:"precision,omitempty"`
+	Min       float64 `json:"min,omitempty"`
+	Max       float64 `json:"max"`
+	Increment float64 `json:"increment"`
+}
+
+// GranularityRange struct defines a range of prices used by PriceGranularity
+type GranularityRange struct {
+	Precision int
+	Min       float64
+	Max       float64
+	Increment float64
+}
+
+// UnmarshalJSON : custom unmarshaller to handle legacy string granularites.
+func (pg *PriceGranularity) UnmarshalJSON(b []byte) error {
+	// We default to medium
+	if len(b) == 0 {
+		*pg = priceGranularityMed
+		return nil
+	}
+	// First check for legacy strings
+	var pgString string
+	err := json.Unmarshal(b, &pgString)
+	if err == nil {
+		*pg = PriceGranularityFromString(pgString)
+		if len(*pg) > 0 {
+			// Only exit if we matched something, else we try processing as custom granularity
+			// This way we error as expecting the new custom granularity standard.
+			return nil
+		}
+	}
+	// Not legacy, so we do a normal Unmarshal
+	gran := []GranularityRange{}
+	err = json.Unmarshal(b, &gran)
+	if err != nil {
+		return err
+	}
+	if len(gran) > 1 {
+		// We only need to loop over the ranges if we have more than one (inter-range checks)
+		precision := gran[0].Precision
+		var prevMax float64 = 0
+		for i, gr := range gran {
+			if gr.Precision != precision {
+				return errors.New("Price granularity error: precision not consistent across entires")
```

Comment:
Is this actually an error? It seems like a legitimate input to me.

Reply:
It gets a little confusing if the precision changes. I think there is a much greater chance that the precision is a typo if it changes than it being a legitimate desire to have the price bucket string change precision in the middle of the range. I am open to changing it, but my feeling is to set this as an expectation, and if someone has a need submit an issue or PR to change it.

Reply:
That's fair. I checked in with Kendall too, and he said he's never seen anyone want to do it either.

In that case, do you think we should change the API? It seems silly for publishers to define it in multiple places if we force it to be consistent anyway. For example, it could be:

```
{
  "precision": 2,
  "ranges": [
    { "min": 0, "max": 0.5, "increment": 0.1 },
    { "min": 0.5, "max": 1.0, "increment": 0.2 },
    ...
  ]
}
```

Reply:
I thought this format was mirroring how we define it in prebid.js. I would want to be consistent in layout with what we do there to reduce confusion. But otherwise I am all for pulling precision out.